### PR TITLE
Integrate serial from upstream rust-vmm 

### PR DESCRIPTION
**Pull request by AlexandruCihodaru**
# Reason for This PR

Use upstream https://github.com/rust-vmm components.
Consume upstream implementation of serial device from rust-vmm.

## Description of Changes

This patch series migrates Firecracker from using local serial implementation, to using rust-vmm/vm-superio/serial
* Defined a wrapper over SerialEvents. This wrapper contains  uart metrics and the buffer_ready file descriptor.
* Defined a wrapper over upstream Serial. This wrapper is needed because one can not implement a trait for a struct defined in another crate. This wrapper also contains the input field that is missing from upstream implementation.
* Implement BUS and EventSubscriber traits for Serial device.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/devices/tests/integration_tests.rs [hunk 1]
```diff
@@ -22,13 +26,18 @@ fn test_issue_serial_hangup_anon_pipe_while_registered_stdin() {
...
+        serial: Serial::with_events(
+            EventFdTrigger::new(EventFd::new(EFD_NONBLOCK).unwrap()),
+            SerialEventsWrapper {
+                metrics: METRICS.uart.clone(),
+                buffer_ready_event_fd: Some(kick_stdin_evt.try_clone().unwrap()),
+            },
+            Box::new(io::stdout()),
+        ),
+        input: Some(Box::new(serial_in)),
+    }));
```
Comment (acatangiu) [inline src/devices/tests/integration_tests.rs:40]: nit: dedupe this code in a helper function

File: src/vmm/src/builder.rs [hunk 2]
```diff
@@ -326,6 +327,9 @@ pub fn build_microvm_for_boot(
     #[allow(unused_mut)]
     let mut boot_cmdline = boot_config.cmdline.clone();
 
+    // Timestamp for measuring microVM boot duration.
+    let request_ts = TimestampUs::default();
+
```
Comment (acatangiu) [inline src/vmm/src/builder.rs:332]: This is a unrelated change (basically reverts an unrelated commit), make sure to drop this change.

File: src/vmm/src/device_manager/legacy.rs [hunk 3]
```diff
@@ -85,27 +89,51 @@ impl PortIODeviceManager {
...
+                    serial: Serial::with_events(
+                        self.com_evt_2_4.try_clone().map_err(Error::EventFd)?,
+                        SerialEventsWrapper {
+                            metrics: METRICS.uart.clone(),
+                            buffer_ready_event_fd: None,
+                        },
+                        Box::new(std::io::sink()),
+                    ),
+                    input: None,
+                })),
```
Comment (acatangiu) [inline src/vmm/src/device_manager/legacy.rs:102]: nit: maybe you can use some temporary variables to simplify this code block. You might even want to dedupe it as it is repeated a few times.

**Review by acatangiu — COMMENTED** [review]
First pass review, another in-depth to follow 👍

**Review by acatangiu — DISMISSED** [review]
LGTM! Are we promoting this to non-draft?

File: src/devices/Cargo.toml [hunk 4]
```diff
@@ -11,7 +11,7 @@ timerfd = ">=1.0"
 versionize = ">=0.1.6"
 versionize_derive = ">=0.1.3"
 vm-memory = { path = "../vm-memory" }
-vm-superio = ">=0.2.0"
+vm-superio = { git = "https://github.com/AlexandruCihodaru/vm-superio.git", branch = "event_trait" }
```
[the remark this answers is not in the document]
Comment (AlexandruCihodaru) [reply src/devices/Cargo.toml]: updated deps.

**Review by acatangiu — APPROVED** [review]

**Review by alindima — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]
